internal/internal_command_state_machine.go

```diff
@@ -284,6 +293,69 @@ func (h *commandsHelper) newCommandStateMachineBase(commandType commandType, id
 	}
 }
 
+// acceptWorkflowUpdate arranges for an AcceptWorkflowUpdate command to be added
+// to the current batch of outgoing commands.
+func (h *commandsHelper) acceptWorkflowUpdate(updateID string) {
+	sm := h.newAcceptWorkflowUpdateStateMachine(updateID)
+	h.addCommand(sm)
+}
+
+// acceptWorkflowUpdate arranges for an CompleteWorkflowUpdate command to be added
+// to the current batch of outgoing commands.
+func (h *commandsHelper) completeWorkflowUpdate(
+	updateID string,
+	success *commonpb.Payloads,
+	failure *failurepb.Failure,
+	durabilityPref enumspb.UpdateWorkflowRejectionDurabilityPreference,
+) {
+	sm := h.newCompleteWorkflowUpdateStateMachine(updateID, success, failure, durabilityPref)
+	h.addCommand(sm)
+}
+
+func (h *commandsHelper) newCompleteWorkflowUpdateStateMachine(
+	updateID string,
+	success *commonpb.Payloads,
+	failure *failurepb.Failure,
+	durabilityPref enumspb.UpdateWorkflowRejectionDurabilityPreference,
+) commandStateMachine {
+	attrs := &commandpb.Command_CompleteWorkflowUpdateCommandAttributes{
+		CompleteWorkflowUpdateCommandAttributes: &commandpb.CompleteWorkflowUpdateCommandAttributes{
+			UpdateToken:          []byte(updateID), // todo: this will change to UpdateId
+			DurabilityPreference: durabilityPref,
+		},
+	}
+	if failure != nil {
+		attrs.CompleteWorkflowUpdateCommandAttributes.Result = &commandpb.CompleteWorkflowUpdateCommandAttributes_Failure{
+			Failure: failure,
+		}
+	} else {
+		attrs.CompleteWorkflowUpdateCommandAttributes.Result = &commandpb.CompleteWorkflowUpdateCommandAttributes_Success{
+			Success: success,
+		}
+	}
```

Comment:
Consider erroring out if both are populated

Reply:
I've been thinking about this but I think I prefer the current logic, here's the reasoning: imagine you're looking at this function in isolation and don't know anything about how things further down the stack are implemented ... this is Go so a function returning (*Foo, error) is common but so to is a function returning (int, error), and in the latter case the trivial coercion to (interface{}, error) will end up with both params being non-nil. So I think the least surprising thing to do here is to discard success value if error is non-nil. Possibly better param names (maybeSuccess, maybeErrror) would help.

In practice FWIW, our `executeFunction` reflection func takes the same approach as above and thus by the time the success/error pair reach this point here you'll have at most one of these two parameters set to something non-nil. However that's not really something that one can encode in Go's type system without getting pretty far away from idiomatic.